hypervisor/arch/x86/guest/vcpu.c

```diff
@@ -641,6 +657,9 @@ void offline_vcpu(struct acrn_vcpu *vcpu)
 	if (vcpu->state == VCPU_ZOMBIE) {
 		vlapic_free(vcpu);
 		per_cpu(ever_run_vcpu, pcpuid_from_vcpu(vcpu)) = NULL;
+
+		per_cpu(vcpu_array, pcpuid_from_vcpu(vcpu))[vcpu->vm->vm_id] = NULL;
```

Comment:
Add a comment: This operation must be atomic to avoid contention with xxxxxxx

Reply:
Not sure if we need to add a comment here, reasons:

1. set setting an unsetting of per_cpu(vcpu_array, pcpu_id)[vm->vm_id] 
in  create_vcpu and offline_vcpu should all be atomic?
per_cpu(vcpu_array, pcpu_id)[vm->vm_id] = vcpu;
per_cpu(vcpu_array, pcpu_id)[vm->vm_id] = NULL;

Also ,create_vcpu and offline_vcpu cannot  be called at same time.

2. I added the following comment in vcpu_handle_post_intr (see below):
/*
 * shutdown_vm would unregister the devices before offline_vcpu is called,
 * so spinlock is not needed to protect access to vcpu_array and vcpu.
 *
 * @pre (vcpu_index < CONFIG_MAX_VM_NUM) && (get_pi_desc(get_cpu_var(vcpu_array)[vcpu_index]) != NULL)
 */
vcpu_handle_post_intr(uint32_t vcpu_index)
{}

Reply:
Change of the field(per_cpu(vcpu_array, pcpu_id)[vm->vm_id] ) must be atomic. Otherwise the WNV handler may see an intermediate value which is invalid.

Reply:
you are using _aligned(8) attribute for the vcpu_array. How about a comment at the place of variable declaration stating that "vcpu_array address must be aligned to 64-bit for atomic access to avoid contention between offline_vcpu and wnv handler"?

Reply:
Yes, I will add this comment

Reply:
OK, then can we add an ASSERT if the alignment is violated?
If we can do that in compile time, that is much better.

Reply:
Do we need to do this? as _aligned(8) would guarantee that the vcpu_array is 64-bit aligned?
We have quite some struct with the aligned(). If we need to do this for vcpu_array, does not this mean we need to also do this for all stuct that used aligned()?